File: SumRootToLeaf/sum_root_to_leaf_numbers.rb
# Definition for a binary tree node.
class TreeNode
    attr_accessor :val, :left, :right
    def initialize(val)
        @val = val
        @left, @right = nil, nil
    end
end

def sum_numbers(root)
    sum_helper(root, "", 0)
end

def sum_helper(n, v, sum)
    return 0 unless n
    unless n.left || n.right
       return sum += ((v+n.val.to_s).to_i)
    end

    lsum = sum_helper(n.left, v + n.val.to_s, sum)
    rsum = sum_helper(n.right, v + n.val.to_s, sum)
    
    lsum + rsum
end

File: SumRootToLeaf/sumRootToLeafNumbers.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def sumNumbers(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if (root is None):
            return 0
        
        numbers = []
        m = [(root, "")]
        while(len(m)!=0):
            node, number = m.pop()
            if (node.left):
                m.append((node.left, number+str(node.val)))
            
            if (node.right):
                m.append((node.right, number+str(node.val)))
            
            if (node.left is None and node.right is None):
                numbers.append(int(number+str(node.val)))

        return sum(numbers)